File: ui/orm/Team.js
import BaseModel from '~/orm/BaseModel'

export default class Team extends BaseModel {
  static entity = 'teams'

  static fields() {
    return {
      id: this.string(),
      role: this.string(),
      number: this.number(),
      beginner: this.boolean(),
      name: this.string(),
      organization: this.string(),
      color: this.string(),
      secretText: this.string().nullable(),
      createdAt: this.string(),
      updatedAt: this.string()
    }
  }

  static mutationFields() {
    return {
      role: '',
      name: '',
      organization: '',
      secretText: '',
      number: 0,
      color: '#FFFFFF',
      beginner: false,
      password: null
    }
  }

  get displayName() {
    return `No.${this.number} ${this.name}`
  }

  get numberStr() {
    return String(this.number)
  }

  get isStaff() {
    return this.role === 'staff'
  }

  get isAudience() {
    return this.role === 'audience'
  }

  get isPlayer() {
    return this.role === 'player'
  }

  static get players() {
    return Team.all().filter(t => t.isPlayer)
  }

  // team99は毎回使われるテストユーザー
  static get playersWithoutTeam99() {
    return Team.all().filter(t => t.isPlayer && t.name !== 'team99')
  }
}

File: api/spec/factories/team.rb
# frozen_string_literal: true

FactoryBot.define do
  alphabets = ('a'..'zz').to_a.freeze

  factory :team do
    name { "team #{alphabets[number - 1]}" }
    password { name }
    organization { "Org. #{name}" }
    secret_text { Random.rand(3).zero? ? Array.new(Random.rand(1..2)) { Faker::Books::Dune.quote }.join("\n") : '' }
    sequence(:number) {|n| n }
    created_at { Time.current - Random.rand(60).minutes - Random.rand(60).seconds }
    updated_at { created_at }
    # before_createだとbulk_insertが失敗するためhas_secure_token 長さ指定を無視してる
    channel { ApplicationRecord.generate_unique_secure_token }

    trait :staff do
      name { "staff #{alphabets[number - 1]}" }
      role { :staff }
      beginner { false }
    end

    trait :audience do
      name { "audience #{alphabets[number - 1]}" }
      role { :audience }
      beginner { false }
    end

    trait :player do
      name { "team #{alphabets[number - 1]}" }
      role { :player }
      beginner { Random.rand(4).zero? } # 1/4ぐらいをbeginnerにする
      color { Random.rand(2).odd? ? nil : Faker::Color.hex_color }
    end

    # association :answers # optional: nil
    # association :attachments # optional: nil
    # association :first_correct_answers # optional: nil
    # association :issues # optional: nil
    # association :notices # optional: nil
    # association :problem_environments # optional: nil
  end
end
